Export jl_resolve_globals_in_ir in Julia src

NHDaly opened the pull request:
This is required (I think) in order for packages to fully implement their own generated-functions macros with custom behavior.

The motivating example for this is https://github.com/NHDaly/StagedFunctions.jl, which provides a complete, working reference implementation for generated functions that are recompiled whenever any of their dependencies change. However, currently it requires modifying Julia to export this function.

(Also, if we expose this function, and then also expose some kind of `jl_apply_in_world_age`, then we could actually move much of the implementation of Julia's generated functions into julia/base .jl files instead of in julia/src .c files!)

NHDaly committed:
Export jl_resolve_globals_in_ir in Julia src

vtjnash commented on src/julia_internal.h:
```diff
@@ -1016,6 +1016,10 @@ void jl_log(int level, jl_value_t *module, jl_value_t *group, jl_value_t *id,
 
+// TODO(NHDALY): Find the right spot for this.
+JL_DLLEXPORT void jl_resolve_globals_in_ir(jl_array_t *stmts, jl_module_t *m,
+                              jl_svec_t *sparam_vals, int binding_effects);
```
this should go in `julia.h` if you want to depend on it. also take it out of `src/toplevel.c`

vtjnash commented on src/method.c:
```diff
@@ -191,8 +191,8 @@ static jl_value_t *resolve_globals(jl_value_t *expr, jl_module_t *module, jl_sve
-void jl_resolve_globals_in_ir(jl_array_t *stmts, jl_module_t *m, jl_svec_t *sparam_vals,
-                              int binding_effects)
+JL_DLLEXPORT void jl_resolve_globals_in_ir(jl_array_t *stmts, jl_module_t *m,
+                              jl_svec_t *sparam_vals, int binding_effects)
```
needs to be realigned (either to constant 8 spaces or floating on the `(`)

NHDaly committed:
Export new `jl_expand_and_resolve`, instead of jl_resolve_globals_in_ir.

Instead of exporting the internal function jl_resolve_globals_in_ir,
this factors out the code from generated functions into
`jl_expand_and_resolve`, which lowers an expression into a CodeInfo and
then resolves globals in the codeinfo in light of provided typenames.

I then exposed this to Julia via another method for `Meta.lower()`, this
one taking type parameters.

NHDaly committed:
Comment `lower_and_resolve_lambda`

NHDaly committed:
Remove the exposed Julia binding for ccall(:jl_expand_and_resolve), because it's very particular about its inputs

NHDaly committed:
Merge branch 'master' into export_jl_resolve_globals_in_ir

NHDaly committed:
Merge branch 'master' into export_jl_resolve_globals_in_ir

NHDaly committed:
Merge branch 'master' into export_jl_resolve_globals_in_ir

NHDaly committed:
Restore error states to what it was before this PR: aways error if !is_code_info, maybe without resetting in_pure_callback

StefanKarpinski commented:
Bump. @vtjnash, can you review?

vtjnash reviewed (approved):
yeah, should be pretty stable

NHDaly commented on src/method.c:
```diff
@@ -376,7 +378,19 @@ STATIC_INLINE jl_value_t *jl_call_staged(jl_method_t *def, jl_value_t *generator
-// return a newly allocated CodeInfo for the function signature
+// Lower `ex` into Julia IR, and (if it expands into a CodeInfo) resolve global-variable
+// references in light of the provided type parameters.
+JL_DLLEXPORT jl_code_info_t *jl_expand_and_resolve(jl_value_t *ex, jl_module_t *module,
```
Before this is merged, I should probably add a comment explaining that on error, this will return an expression with `error_sym` in its head.

NHDaly replied:
Done.

NHDaly committed:
Add error return value semantics to comment on jl_expand_and_resolve

NHDaly commented:
Okay, thanks for the review! :) I've updated the comments to my satisfaction, so I'm happy to have this merged whenever! Thanks @vtjnash!

vchuravy commented:
AnalyzeGC failure is real

NHDaly commented:
woah, oops. Thanks Valentin!

I've not done too much of this stuff.. I'll read up on the docs and figure out how i'm supposed to fix this. :) ❤️ thanks!

NHDaly committed:
Add GC annotations inside `jl_expand_and_resolve`

(I think I did this right? I couldn't get the static analysis to build
on my machine...)

NHDaly commented:
Cool, thanks @vchuravy! Fixed the test. (I think the `test_win32` is a flake?)

c42f commented:
> and then also expose some kind of `jl_apply_in_world_age`

Cool. FWIW this was implemented in #35844

NHDaly commented:
> > (Also, if we expose this function, and then also expose some kind of `jl_apply_in_world_age`, then we could actually move much of the implementation of Julia's generated functions into julia/base .jl files instead of in julia/src .c files!)

> > and then also expose some kind of `jl_apply_in_world_age`
> 
> Cool. FWIW this was implemented in #35844

Oooh, very cool indeed! It might be fun to do that rewrite at some point! 😁
